src/IssueViz/Editor/SelectedIssueTagging/Adornment/IssueLocationAdornmentTagger.cs

```diff
@@ -18,20 +18,46 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Text;
 using Microsoft.VisualStudio.Text.Editor;
 using Microsoft.VisualStudio.Text.Tagging;
+using SonarLint.VisualStudio.IssueVisualization.Models;
 
 namespace SonarLint.VisualStudio.IssueVisualization.Editor.SelectedIssueTagging.Adornment
 {
     internal sealed class IssueLocationAdornmentTagger : FilteringTaggerBase<ISelectedIssueLocationTag, IntraTextAdornmentTag>
     {
         private readonly IWpfTextView wpfView;
+        private readonly ICachingAdornmentFactory adornmentCache;
+        private readonly ITagAggregator<ISelectedIssueLocationTag> tagAggregator;
 
         public IssueLocationAdornmentTagger(ITagAggregator<ISelectedIssueLocationTag> tagAggregator, IWpfTextView textView)
-            : base(tagAggregator, textView)
+            : this(tagAggregator, textView, new CachingAdornmentFactory(textView))
         {
-            wpfView = textView;
+        }
+
+        internal /* for testing */ IssueLocationAdornmentTagger(ITagAggregator<ISelectedIssueLocationTag> tagAggregator,
+            IWpfTextView wpfView, ICachingAdornmentFactory adornmentCache)
+            : base(tagAggregator, wpfView)
+        {
+            this.wpfView = wpfView;
+            this.adornmentCache = adornmentCache;
+
+            this.tagAggregator = tagAggregator;
+            tagAggregator.BatchedTagsChanged += OnTagsChanged;
```

Comment:
It feels wrong having to register to the event twice in the same class instance. Perhaps we can make the base `NotifyTagsChanged` a virtual, but I don't know if it's better.

Reply:
There are arguments both ways.  We could just make `OnAggregatorTagsChanged` virtual, but we don't want subclasses to be able to override the base class behaviour (or rely on them calling the base class method). We could add another empty virtual method as an extension point that is called by the base class to all subclasses to extend if necessary, but at this point that would be adding an extension point specifically to handle the adornment tagger case. Or we have the adornment tagger adds its own functionality separately by registering to the aggregator event.